src/System.Collections/src/System/Collections/Generic/HashSet.cs

```diff
@@ -1711,6 +1716,95 @@ internal static bool HashSetEquals(HashSet<T> set1, HashSet<T> set2, IEqualityCo
+        /// to specified comparer.
+        /// Because items are hashed according to a specific equality comparer, we have to resort
+        /// to building a new set if neither use that equality comparer.
+        /// </summary>
+        /// <param name="set1">The first set to compare.</param>
+        /// <param name="set2">The second set to compare.</param>
+        /// <param name="comparer">The comparer for comparing elements.</param>
+        /// <returns><c>true</c> if the two sets are equivalent, <c>false</c> otherwise.</returns>
+        internal static bool HashSetEquals(HashSet<T> set1, HashSet<T> set2, IEqualityComparer<T> comparer)
+        {
+            if (set1 == set2)
+            {
+                return true;
+            }
+
+            if (set1 == null | set2 == null)
+            {
+                return false;
+            }
+
+            int count1 = set1.Count;
+            int count2 = set2.Count;
+            if (count1 == 0)
+            {
+                return count2 == 0;
+            }
+
+            if (count2 == 0)
+            {
+                return false;
+            }
+
+            if (comparer.Equals(set1._comparer))
+            {
+                if (comparer.Equals(set2._comparer))
+                {
+                    // Since both sets are using our comparer as their comparer, we can test for set-equality.
+                    return set1.Count == set2.Count && set1.ContainsAllElements(set2);
+                }
+
+                if (count1 > count2)
+                {
+                    // Filtering set2 according to comparer can result in fewer than set2.Count elements,
+                    // but never more, so short-circuit if set2.Count is less than set1.Count.
+                    return false;
+                }
+            }
+            else
+            {
+                // We need a starting point of a set according to the comparer, and set1 doesn't meet
+                // that criteria. If set2 does, swap to use it.
+                if (comparer.Equals(set2._comparer))
+                {
+                    if (count1 < count2)
+                    {
+                        return false;
+                    }
+
+                    HashSet<T> swapTemp = set1;
+                    set1 = set2;
+                    set2 = swapTemp;
+                    count1 = count2;
+                }
+                else
+                {
+                    // Niether set use our comparer, so create a new one with the contents of set1.
+                    set1 = new HashSet<T>(set1, comparer);
+
+                    if ((count1 = set1.Count) > count2)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            HashSet<T> seen = new HashSet<T>(count1 + 1, comparer);
+            foreach (T item in set2)
+            {
+                if (seen.Add(item) && !set1.Contains(item))
```

Comment:
The old algorithm would show them as equal or not equal depending on which order they were passed to it.

The main use-case I can see would be for cases where the comparer was the same as those in the sets, but it should also be consistent in that whether the comparer was the same or not:

1. `Equals(a, b) == Equals(b, a)` (not true with the first overload)
2. If `Equals(a, b)` then always `GetHashCode(a) == GetHashCode(b)` (not true with the first overload)
3. People for whom the old code works aren't broken by a change.
4. We don't have just a new public method with the same signature and different name and behaviour (https://github.com/dotnet/corefx/issues/12560#issuecomment-262080511 and https://github.com/dotnet/corefx/issues/12560#issuecomment-262085527)

If we don't treat cases like the above where a set hasn't the comparer used as its comparer by forcing comparison to use that comparer then points 1 and 2 can't stand (we can't know which comparer to use for `GetHashCode` if we don't know which will be used for `Equals`) unless perhaps we consider any such set being passed in to be an `ArgumentException` and refuse to do the `Equals` at all.